### PR TITLE
Dockerfile with cache

### PR DESCRIPTION
**Pull request by michaelneale**
this shows how caching works

### CODE REVIEW

File: Dockerfile.build [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+COPY pom.xml ./
+
+# Run install task so all necessary dependencies are downloaded and cached in
+# the Docker image. We're running through the whole process but disable
+# testing and make sure the command doesn't fail.
+RUN mvn install clean --fail-never -B -DfailIfNoTests=false
+
+
+# This gets the source in place for codeship to do its thing from codeship-steps.yml
+COPY . ./
```
Comment (oleg-nenashev) [inline Dockerfile.build:13]: You do not really use it, code is passed through the workspace

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
It is not really a dependency cache, it is just an image build cache. In the current implementation it will partially do the job, when pom.xml is not modified. Any pom.xml modification will lead to the dependency layer rebuild. Better than nothing, but should be advertised with proper disclaimer IMHO.

Requesting changes because of `COPY . ./` which is not actually used in the build.

File: Dockerfile.build (hunk 1 above)
Comment (michaelneale) [reply Dockerfile.build:13]: ok will get rid of it - was from an old version of mine. 

I think it is a pretty good cache - a permanent desktop like cache experience is not really ideal IMO (often you end up wanting to blow it away), or there would be a cache *server* somewhere to keep build instances ephemeral. I think it is a reasonable cloud native-like approach (and I think outside of the java ecosystem, not an uncommon way to work, eg in npm there is mostly a per project cache).

Comment (michaelneale) [reply Dockerfile.build:13]: so I think its more semantics - its fine just to say it is a cache (if it was a zip file in s3, similar challenges would result). It should be clear it isn't a cache server though

**Comment by michaelneale** [comment]
I'll put this in the quickstart docs I think

**Pull request closed without merging**